src/rgw/rgw_notify.h

```diff
@@ -43,23 +43,67 @@ int remove_persistent_topic(const std::string& topic_name, optional_yield y);
 // then used to commit or abort the reservation
 struct reservation_t {
   struct topic_t {
-    topic_t(const std::string& _configurationId, const rgw_pubsub_topic& _cfg, cls_2pc_reservation::id_t _res_id) :
-        configurationId(_configurationId), cfg(_cfg), res_id(_res_id) {}
+    topic_t(std::string& _configurationId, rgw_pubsub_topic& _cfg,
+	    cls_2pc_reservation::id_t _res_id) :
+      configurationId(_configurationId), cfg(_cfg), res_id(_res_id) {}
 
-    const std::string configurationId;
-    const rgw_pubsub_topic cfg;
+    std::string configurationId;
+    rgw_pubsub_topic cfg;
     // res_id is reset after topic is committed/aborted
     cls_2pc_reservation::id_t res_id;
   };
-
+  
   std::vector<topic_t> topics;
-  rgw::sal::RGWRadosStore* const store;
-  const req_state* const s;
+  rgw::sal::RGWRadosStore* store;
   size_t size;
-  rgw::sal::RGWObject* const object;
 
-  reservation_t(rgw::sal::RGWRadosStore* _store, const req_state* _s, rgw::sal::RGWObject* _object) : 
-      store(_store), s(_s), object(_object) {}
+  const DoutPrefixProvider* dpp;
+  RGWObjectCtx* obj_ctx;
+  rgw::sal::RGWObject* object;
```

Comment:
So, all this is going to conflict with zipper 10, which wraps reservation_t.  Do we want to hold this until zipper 10 goes in and re-work, or do we want to hold zipper 10 until this goes in and re-work?

Reply:
What do you think will be easier?

Reply:
Probably better to do zipper 10 first.

Reply:
I guess thinking about it, I'm unclear why reservation_t would be subsumed under zipper, though.  RADOS is currently used as a queuing back-end, but even that could at some point be plugged out.  The notification itself is platform agnostic.

Reply:
reservation_t itself is more or less okay, but only because it's not a real class.  publish_reserve() and publish_commit() are rados-specific, and so need a RGWRadosStore pointer, which is provided by reservation_t.   If those methods were actually part of reservation_t, it would also be rados-specific.

Reply:
it seems like we would ideally want these apis to be platform agnostic, but have the reservation step be specialized for rados;  before yuval added persistent queues, I think there was no rados dependency

Reply:
(unlike pub-sub originally, which was rados-specific)

Reply:
zipper 10 provides a platform agnostic version of these APIs here: https://github.com/dang/ceph/blob/wip-dang-zipper-10/src/rgw/rgw_sal_rados.h#L526
I'm open to a better solution, if there is one.

Reply:
we can abstract the queue API - originally we didn't think it has to be a RADOS backed queue (thought of using some shared memory queue for better perf with less persistency guarantees).